enhancements/ingress/mutable-publishing-scope.md

```diff
@@ -0,0 +1,348 @@
+load-balancer between internal and external without deleting and recreating the
+load balancer, by setting an annotation on the Kubernetes Service object.  On
+these platforms, the operator merely sets the annotation to the desired scope,
+and the operation of changing it scope is complete.
+
+Other platforms (such as AWS) require deleting and recreating a load balancer to
+change its scope.  This operation is disruptive: It interrupts ingress traffic
+and may cause the load balancer's address to change.  On these platforms, the
+operator signals that the user must delete the Kubernetes Service object.  Once
+the user performs this step, the operator recreates the load balancer with the
+desired scope to complete the operation.
+
+In order to signal that the user must take action, the operator sets the
+`Progressing` status condition to `True` with a message that indicates the
+action that the user must take:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: custom
+  namespace: openshift-ingress-operator
+spec:
+  replicas: 2
+  domain: mycluster.com
+  endpointPublishingStrategy:
+    type: LoadBalancerService
+    loadBalancer:
+      scope: Internal
+status:
+  availableReplicas: 2
+  conditions:
+  - message: Deployment has minimum availability.
+    reason: MinimumReplicasAvailable
+    status: "True"
+    type: Available
+  - message: Have load balancer with scope "External", want load balancer with scope "Internal".  You can delete the openshift-ingress/router-default service to proceed [...].  Alternatively, you can change the IngressController's spec.endpointPublishingStrategy.loadBalancer.scope field value back to its previous value [...].
+    reason: ScopeChanged
+    status: "True"
+    type: Progressing
+```
+
+Note that the IngressController remains available, and that the user can revert
+the change by setting `spec.endpointPublishingStrategy.loadBalancer.scope` back
+to its previous value.  That means the user must take one of two actions:
+
+1. Delete the Service referenced in the status condition.
+2. Revert the change to the IngressController.
+
+If the user deletes the Service, the operator recreates it with the desired
+scope.  Alternatively, if the user reverts the change, the `Progressing` status
+condition changes back to `False`, and the state of the IngressController is as
+it was before the user first changed the value of
+`spec.endpointPublishingStrategy.loadBalancer.scope`.
+
+In addition to deleting the IngressController explicitly, it is possible to
+annotate it with the newly defined
+`ingress.operator.openshift.io/auto-delete-load-balancer` annotation.  If the
+operator observes that this annotation is set and that its
+`spec.endpointPublishingStrategy.loadBalancer.scope` field has been changed, the
+operator automatically deletes the Service if needed to complete a scope-change
+operation.  This purpose of this annotation is to simplify automation: A tool
```

Comment:
The annotation is sort of awkward. Seems like it would be better specifiied as a flag on the `oc edit` operation, something like `oc edit ... --auto-update` or `--propagate` or similar.

Reply:
Would the `oc patch` command have custom logic to add the `ingress.operator.openshift.io/auto-delete-load-balancer` annotation when the user provided the `--auto-update` command and an IngressController?  I'd rather avoid adding IngressController-specific or (OpenShift-specific) logic to `oc patch`, and anyway, the annotation is intended for automation such as [the cloud-ingress-operator's publishingstrategy controller](https://github.com/openshift/cloud-ingress-operator/blob/f560b7888c0e140cdc4fa4864929638e9982d877/pkg/controller/publishingstrategy/publishingstrategy_controller.go).

Reply:
ok